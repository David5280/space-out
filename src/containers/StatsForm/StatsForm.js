import React from 'react';
import { NavLink } from 'react-router-dom';
import { addUser } from '../../actions';
import { connect } from 'react-redux';

class StatsForm extends React.Component {
  constructor() {
    super()
    this.state = {
      name: '',
      age: '',
      weight: ''
    }
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value })
  }

  handleSubmit = (e) => {
    e.preventDefault();
    this.props.addUser(this.state);
    this.clearState();
  }

  clearState = () => {
    this.setState({
      name: '',
      age: '',
      weight: ''
    })
  }

  render() {
    const planetOptions = this.props.planets.map(planet => {
        return (
          <div 
            className='stats-link-option'
            key={planet.title}
          >
            <NavLink 
              to={`/your-stats/${planet.title}`} 
              >
              {planet.title}
            <img 
              src={planet.cutout} 
              alt={`${planet.title}-cutout`} 
              className='stats-cutout' 
            />
            </NavLink>
          </div>
        )
      })
    return(
      <section className='form-container'>
        {
          (!this.props.user.name &&
            !this.props.user.age &&
            !this.props.user.weight) &&
            <form className='stats-form'>
            <input 
              className='stats-form-input' 
              name='name' 
              placeholder='First Name...'
              type='text'
              onChange={this.handleChange}
              value={this.state.name} 
            />
            <input 
              className='stats-form-input' 
              name='age' 
              placeholder='Age...' 
              type='text'
              onChange={this.handleChange}
              value={this.state.age} 
            />
            <input 
              className='stats-form-input' 
              name='weight'
              placeholder='Weight...  (lbs)' 
              type='text'
              onChange={this.handleChange}
              value={this.state.weight} 
            />
            <button 
              className='stats-form-input'
              onClick={this.handleSubmit}>Submit!</button>
          </form>
        }
        {
          (this.props.user.name &&
            this.props.user.age &&
            this.props.user.weight) &&
          <section className='stats-greeting'>
            <h3>Hello, {this.props.user.name}!</h3>
            <p>Pick a planet below to simulate yourself there!</p>
            <div className='stats-planets-list'>
              {planetOptions}
            </div>
          </section>
        }
      </section>
    )
  }
}

const mapStateToProps = (state) => ({
  user: state.user,
  planets: state.planets
})

const mapDispatchToProps = (dispatch) => ({
  addUser:(user) => dispatch(addUser(user))
})

export default connect(mapStateToProps, mapDispatchToProps)(StatsForm);