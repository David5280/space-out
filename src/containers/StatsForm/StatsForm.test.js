import React from 'react';
import { StatsForm } from './StatsForm';
import { shallow } from 'enzyme';

describe('StatsForm', () => {
  let e;
  let wrapper;
  let mockPlanets;
  let mockUser;
  let mockUser2;
  beforeEach(() => {
    e = {
      target: {
        name: 'name',
        value: 'something'
      }
    };
    mockPlanets = [
      {title: 'earth', population: 'too many'},
      {title: 'mars', population: 'none, yet'}
    ];
    mockUser = {
      name: 'David',
      age: 25,
      weight: '180'
    };
    mockUser2 = {
      name: '',
      age: null,
      weight: ''
    };
    wrapper = shallow(<StatsForm planets={mockPlanets} user={mockUser}/>)

  })
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should change state when handleChange is invoked', () => {
    wrapper.instance().handleChange(e)
    expect(wrapper.state('name')).toEqual('something')
  });
  it('should invoke handleChange on input', () => {
    wrapper = shallow(<StatsForm planets={mockPlanets} user={mockUser2}/>);
    
    wrapper.instance().handleChange = jest.fn();
    wrapper.find('.name-input').simulate('change', e);
    expect(wrapper.instance().handleChange).toHaveBeenCalled();
  });
  it('should call handleSubmit on button click', () => {
    wrapper = shallow(<StatsForm planets={mockPlanets} user={mockUser2}/>);
    wrapper.instance().handleSubmit = jest.fn();
    wrapper.find('.submit-btn').simulate('click');
    expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
  });
});