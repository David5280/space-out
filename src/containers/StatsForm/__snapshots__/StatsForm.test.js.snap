// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsForm should match the snapshot 1`] = `ShallowWrapper {}`;
